### PR TITLE
separate trace history out to a new class

### PR DESCRIPTION
**Pull request by richorama**
I also updated the benchmarks to include profiling of the history.

These are my results:

```
                                Method | SiloCount | GrainTypeCount | GrainMethodCount | HistorySize |      Mean |      Error |     StdDev |
-------------------------------------- |---------- |--------------- |----------------- |------------ |----------:|-----------:|-----------:|
            Test_QueryAll_TraceHistory |        10 |             50 |               10 |         100 | 504.63 ms | 13.6696 ms | 20.8750 ms |
           Test_QuerySilo_TraceHistory |        10 |             50 |               10 |         100 |  63.53 ms |  0.8962 ms |  0.7945 ms |
          Test_QueryGrain_TraceHistory |        10 |             50 |               10 |         100 |  23.75 ms |  0.4648 ms |  0.6044 ms |
 Test_GroupByGrainAndSilo_TraceHistory |        10 |             50 |               10 |         100 | 104.60 ms |  1.4846 ms |  1.1591 ms |
```

### CODE REVIEW

File: OrleansDashboard/History/TraceHistory.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+        }
+
+        public IEnumerable<TraceAggregate> GroupByGrainAndSilo()
+        {
+            return this.history.GroupBy(x => (x.Grain, x.SiloAddress)).Select(x => {
+                return new TraceAggregate{
+                    SiloAddress = x.Key.SiloAddress,
+                    Grain = x.Key.Grain,
+                    Count = x.Sum(y => y.Count),
```
Comment (SebastianStehle) [inline OrleansDashboard/History/TraceHistory.cs]: I know it is a micro-optimization, but a manual loop would be a little bit faster.

File: OrleansDashboard/History/ITraceHistory.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        public long Count { get; set; }
+        public long ExceptionCount { get; set; }
+        public double ElapsedTime { get; set; }
+    }
+
+    public interface ITraceHistory
+    {
+
+        // Add trace information recorded at a silo
+        void Add(DateTime time, string siloAddress, SiloGrainTraceEntry[] grainTrace);
```
Comment (SebastianStehle) [inline OrleansDashboard/History/ITraceHistory.cs]: Why are you not using <summary>? At least you get intellisense for free.

File: OrleansDashboard/History/TraceHistory.cs (hunk 1 above)
Comment (richorama) [reply OrleansDashboard/History/TraceHistory.cs]: good point, switched to a loop and benchmarking.

The 'higher order' programming part of my brain is arguing about it with the performance part :¬)

**Comment by richorama** [comment]
No major difference in performance, but I'll leave it in :¬)

```
                                Method | SiloCount | GrainTypeCount | GrainMethodCount | HistorySize |      Mean |      Error |    StdDev |    Median |
-------------------------------------- |---------- |--------------- |----------------- |------------ |----------:|-----------:|----------:|----------:|
            Test_QueryAll_TraceHistory |        10 |             50 |               10 |         100 | 521.18 ms | 11.4811 ms | 33.309 ms | 512.48 ms |
           Test_QuerySilo_TraceHistory |        10 |             50 |               10 |         100 |  69.80 ms |  2.0933 ms |  5.938 ms |  67.36 ms |
          Test_QueryGrain_TraceHistory |        10 |             50 |               10 |         100 |  27.36 ms |  0.8026 ms |  2.264 ms |  26.90 ms |
 Test_GroupByGrainAndSilo_TraceHistory |        10 |             50 |               10 |         100 | 100.56 ms |  2.3469 ms |  6.658 ms |  98.97 ms |
```

**Pull request merged**